services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,514 @@
+     * @throws NullPointerException if the mongoDbCallExecutor is {@code null}.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = Objects.requireNonNull(mongoDbCallExecutor);
+        this.mongoClient = Objects.requireNonNull(this.mongoDbCallExecutor.getMongoClient());
+    }
+
+    /**
+     * Sets the configuration properties for this service.
+     *
+     * @param config The configuration properties.
+     * @throws NullPointerException if the configuration is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final MongoDbBasedTenantsConfigProperties config) {
+        this.config = Objects.requireNonNull(config);
+    }
+
+    @Override
+    public void start(final Promise<Void> startPromise) {
+        mongoDbCallExecutor.createCollectionIndex(config.getCollectionName(),
+                new JsonObject().put(RegistrationConstants.FIELD_PAYLOAD_TENANT_ID,
+                        1),
+                new IndexOptions().unique(true), INDEX_CREATION_MAX_RETRIES)
+                .map(success -> {
+                    startPromise.complete();
+                    return null;
+                })
+                .onFailure(startPromise::fail);
+    }
+
+    @Override
+    public void stop(final Promise<Void> stopPromise) {
+        this.mongoClient.close();
+        stopPromise.complete();
+    }
+
+    /**
+     * {@inheritDoc}
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>204 No Content</em> if the tenant has been updated successfully.</li>
+     *         <li><em>403 forbidden</em> if tenant modification is forbidden due to configuration settings</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier and version exists.</li>
+     *         <li><em>409 Conflict</em> if provided trusted certificate authority is already present in another
+     *         tenant.</li>
+     *         <li><em>412 Precondition Failed</em> if the version exists but mismatch the current tenants version.</li>
+     *         </ul>
+     */
+    @Override
+    public Future<OperationResult<Void>> updateTenant(final String tenantId, final Tenant tenantObj,
+            final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantId)
+                .compose(ok -> findTenant(tenantId))
+                .compose(tenantDto -> checkConflictingCATenants(Optional.of(tenantDto),
+                        tenantObj.getTrustedCertificateAuthoritySubjectDNs(), span))
+                .compose(tenantDto -> processUpdateTenant(tenantId, tenantObj, resourceVersion, span))
```

Comment:
Finding a tenant, checking for conflict CAs and then updating that tenant are not atomic. After checking and before update some other request could have inserted a tenant with the same _SubjectDN_. So it is better if we these operations atomic.

we could created an index as below. That ensures no conflicting CAs are updated and the code regarding conflicting CAs could be removed. 
``` 
db.getCollection('tenants')
.createIndex(
{"tenant.trusted-ca.subject-dn": 1}, 
{unique: true, partialFilterExpression: { "tenant.trusted-ca": { $exists: true } } })
```

Reply:
Good idea. Done